### PR TITLE
[BEAM-10475] Add max buffering duration option for GroupIntoBatches transform in Python

### PR DESCRIPTION
**Pull request by nehsyc**
The option was added for Java in https://github.com/apache/beam/pull/12726

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by nehsyc** [comment]
R: @boyuanzz

**Comment by nehsyc** [comment]
CC: @robertwb

File: sdks/python/apache_beam/transforms/util.py [hunk 1]
```diff
@@ -777,33 +795,56 @@ def process(
...
+    @on_timer(WINDOW_TIMER)
+    def on_window_timer(
         self,
         element_state=DoFn.StateParam(ELEMENT_STATE),
-        count_state=DoFn.StateParam(COUNT_STATE)):
+        count_state=DoFn.StateParam(COUNT_STATE),
+        buffering_timer=DoFn.TimerParam(BUFFERING_TIMER)):
+      batch = [element for element in element_state.read()]
+      if batch:
+        key, _ = batch[0]
```
Comment (robertwb) [inline sdks/python/apache_beam/transforms/util.py:827]: Pull this (and the exact same logic above and below) into common helper method.

File: sdks/python/apache_beam/transforms/util.py [hunk 2]
```diff
@@ -1009,7 +1050,7 @@ def _process(element):
 
   @staticmethod
   @typehints.with_input_types(str)
-  @typehints.with_output_types(Union[List[str], List[Tuple[str, str]]])
+  @typehints.with_output_types(Union[List[str], Tuple[str, str]])
```
Comment (robertwb) [inline sdks/python/apache_beam/transforms/util.py:1053]: Was this wrong?

File: sdks/python/apache_beam/transforms/util_test.py [hunk 3]
```diff
@@ -110,10 +117,16 @@ def test_windowed_batches(self):
...
-          5, 5, 10, 10,  # elements in [0, 30)
-          10, 7,         # elements in [30, 47)
-      ]))
+      assert_that(
+          res,
+          equal_to([
+              5,
+              5,
+              10,
+              10,  # elements in [0, 30)
```
Comment (robertwb) [inline sdks/python/apache_beam/transforms/util_test.py:126]: Nit: I prefer the previous formatting (as the comment applies to the whole set of elements).

File: sdks/python/apache_beam/transforms/util_test.py [hunk 4]
```diff
@@ -644,6 +657,12 @@ def _create_test_data():
       data.append(("key", scientists[index]))
     return data
 
+  class _ExpandIterable(DoFn):
```
Comment (robertwb) [inline sdks/python/apache_beam/transforms/util_test.py:660]: _ExpandValuesIterable? 

This could also be `FlatMapTuple(lambda k, vs: vs)`

**Review by robertwb — COMMENTED** [review]
Thanks. Just some minor comments.

File: sdks/python/apache_beam/transforms/util.py (hunk 1 above)
Comment (nehsyc) [reply sdks/python/apache_beam/transforms/util.py]: Done.

File: sdks/python/apache_beam/transforms/util.py (hunk 2 above)
Comment (nehsyc) [reply sdks/python/apache_beam/transforms/util.py]: oops. It was correct. Reverted.

File: sdks/python/apache_beam/transforms/util_test.py (hunk 3 above)
Comment (nehsyc) [reply sdks/python/apache_beam/transforms/util_test.py]: Reverted.

File: sdks/python/apache_beam/transforms/util_test.py (hunk 4 above)
Comment (nehsyc) [reply sdks/python/apache_beam/transforms/util_test.py]: Good to know! Replaced with `FlatMapTuple` :D

**Review by boyuanzz — COMMENTED** [review]
Left one comment. Other than that, it looks good to me.

File: sdks/python/apache_beam/transforms/util.py [hunk 5]
```diff
@@ -751,24 +751,42 @@ class GroupIntoBatches(PTransform):
...
+      max_buffering_duration_secs: (optional) How long in seconds at most an
+        incomplete batch of elements is allowed to be buffered in the states.
+        The duration must be a positive second duration and should be given as
+        an int or float.
+      clock: (optional) an alternative to time.time (mostly for testing)
     """
     self.batch_size = batch_size
 
+    if max_buffering_duration_secs is not None:
+      assert max_buffering_duration_secs > 0, \
```
[the remark this answers is not in the document]
Comment (nehsyc) [reply sdks/python/apache_beam/transforms/util.py]: Done.

**Comment by nehsyc** [comment]
> Left one comment. Other than that, it looks good to me.

Thanks for the review! I squashed the commits into one.

File: sdks/python/apache_beam/transforms/util.py [hunk 6]
```diff
@@ -777,33 +795,56 @@ def process(
...
-        batch_values = [v for (k, v) in batch]
-        yield (key, batch_values)
-        element_state.clear()
-        count_state.clear()
-
-    @on_timer(EXPIRY_TIMER)
-    def expiry(
+        key, batch_values = self.flush_batch(
+            element_state, count_state, buffering_timer)
+        if key is not None:
```
Comment (robertwb) [inline sdks/python/apache_beam/transforms/util.py:813]: None is a valid key, so this is not quite safe. Instead, you could make flush_batch an iterable, and just do `return self.flush_batch()` everywhere.

File: sdks/python/apache_beam/transforms/util.py [hunk 7]
```diff
@@ -777,33 +795,56 @@ def process(
...
+    def flush_batch(self, element_state, count_state, buffering_timer):
       batch = [element for element in element_state.read()]
-      if batch:
-        key, _ = batch[0]
-        batch_values = [v for (k, v) in batch]
-        yield (key, batch_values)
-        element_state.clear()
-        count_state.clear()
+      if not batch:
+        return None, None
```
Comment (robertwb) [inline sdks/python/apache_beam/transforms/util.py:841]: Specifically, here you would just `return` and below you would `yield key, batch_values`.

**Review by robertwb — COMMENTED** [review]
Thanks. Just one more comment.

File: sdks/python/apache_beam/transforms/util.py (hunk 6 above)
Comment (nehsyc) [reply sdks/python/apache_beam/transforms/util.py]: Done

File: sdks/python/apache_beam/transforms/util.py (hunk 7 above)
Comment (nehsyc) [reply sdks/python/apache_beam/transforms/util.py]: Gotcha. I missed the return case previously so it didn't work correctly. Thanks for the suggestion!

**Comment by nehsyc** [comment]
Run PreCommit

**Comment by boyuanzz** [comment]
Run Python PreCommit

**Comment by nehsyc** [comment]
> Run Python PreCommit

The PreCommit failure was related. I pushed a fix to direct_runner.py.

**Comment by boyuanzz** [comment]
Run Python_PVR_Flink PreCommit

**Comment by boyuanzz** [comment]
Per https://issues.apache.org/jira/browse/BEAM-10921, we can ignore the windows-latest test suite.

**Review by boyuanzz — APPROVED** [review]
Thanks!

**Pull request merged**